### PR TITLE
gp by reference fixes

### LINKED ISSUE #317: gp_QuaternionSLerp doesnt return changed by-ref argument
```python
from OCC.gp import gp_Vec, gp_Quaternion, gp_QuaternionSLerp

vX = gp_Vec(12, 0, 0)
vY = gp_Vec(0, 12, 0)
v45 = (gp_Vec(1, 1, 1).Normalized() * 12)

q = gp_Quaternion()
q1 = gp_Quaternion(vX, vX)
q2 = gp_Quaternion(vX, vY)

interp = gp_QuaternionSLerp(q1, q2)

for i in range(100):
    i__ = i / 100.
    interp.Interpolate(i__, q)
    print("i: ", i__)
    print("quat X,Y,Z,W: {} / {} / {} / {}".format(q.X(), q.Y(), q.Z(), q.W()))


```

---

quat X,Y,Z,W: 0.0 / 0.0 / 0.0 / 1.0
i:  0.96
quat X,Y,Z,W: 0.0 / 0.0 / 0.0 / 1.0
i:  0.97
quat X,Y,Z,W: 0.0 / 0.0 / 0.0 / 1.0
i:  0.98
quat X,Y,Z,W: 0.0 / 0.0 / 0.0 / 1.0
i:  0.99
quat X,Y,Z,W: 0.0 / 0.0 / 0.0 / 1.0

looks like the by-ref argument `q` is not updated?

### PR DESCRIPTION
**Pull request by tpaviot**
This branch adds back by ref passing/return in the gp module. gp is now considered as other modules.

Fixes #317

### CODE REVIEW

File: examples/core_geometry_quaternion.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    for i in frange(0, 1.0, 0.01):
+        interp.Interpolate(i, q)
+        # displace the white edges a little from the origin so not to obstruct the other edges
+        v = gp_Vec(0, -24*i, 0)
+        q_v_ = q * v45
+
+
+        print("q: {} / {} / {} / {}".format(q.X(), q.Y(), q.Z(), q.W()))
+        print("vvvv {} / {} / {}".format(q_v_.X(), q_v_.Y(), q_v_.Z()))
```
Comment (jf---) [inline examples/core_geometry_quaternion.py]: can be removed

File: examples/core_geometry_quaternion.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+        p = gp_Pnt((q_v_ + v).XYZ())
+        v__as_pnt = gp_Pnt((origin + v).XYZ())
+        e = make_edge(v__as_pnt, p)
+        display.DisplayColoredShape(e, 'WHITE')
+        msg = 'v45->q1*v45 @{0}'.format(i / 10.)
```
Comment (jf---) [inline examples/core_geometry_quaternion.py]: same here

File: examples/core_geometry_quaternion.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+
+@wraps(BRepBuilderAPI_MakeEdge)
+def make_edge(*args):
+    edge = BRepBuilderAPI_MakeEdge(*args)
+    with assert_isdone(edge, 'failed to produce edge'):
+        result = edge.Edge()
+        edge.Delete()
+        return result
+
```
Comment (jf---) [inline examples/core_geometry_quaternion.py]: should use `make_edge` from `core_geometry_utils`, perhaps `make_box` should go ther too?

File: examples/core_geometry_quaternion.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+from OCC.BRepBuilderAPI import BRepBuilderAPI_MakeEdge
+from OCC.BRepPrimAPI import BRepPrimAPI_MakeBox
+from OCC.Display.SimpleGui import init_display
+from OCC.gp import gp_QuaternionSLerp, gp_Quaternion, gp_Vec, gp_Pnt
+
+display, start_display, add_menu, add_function_to_menu = init_display()
+
+
+class assert_isdone(object):
```
Comment (jf---) [inline examples/core_geometry_quaternion.py]: if used, should be at `core_geometry_utils`

File: examples/core_geometry_quaternion.py (hunk 1 above)
Comment (tpaviot) [inline examples/core_geometry_quaternion.py]: done

File: examples/core_geometry_quaternion.py (hunk 2 above)
Comment (tpaviot) [inline examples/core_geometry_quaternion.py]: done

File: examples/core_geometry_quaternion.py (hunk 4 above)
Comment (tpaviot) [inline examples/core_geometry_quaternion.py]: done

File: examples/core_geometry_quaternion.py (hunk 3 above)
Comment (tpaviot) [inline examples/core_geometry_quaternion.py]: not necessary

**Pull request merged**